### PR TITLE
Enable JITServer in CRIU non-portable mode

### PR DESCRIPTION
**Pull request by ymanton**
To make life simpler for end users CRIU mode can also implicitly enable JITServer client mode so that restored processes can seamlessly connect to a JITServer should one be available.

This PR accomplishes two things:

1. Implicitly enables JITServer when CRIU non-portable mode is enabled (and allows disabling via explicit `-XX:-UseJITServer`)
2. Delays remote compilations in non-portable mode until after the process is restored

Neither of the above happens in portable mode. Users can explicitly enable JITServer in portable mode via `-XX:+UseJITServer` if they wish. Delaying compilations until after restore doesn't make sense in portable mode since there is no limit on how many times the process can be checkpointed/restored.

Here's a summary of operations under various combinations of options:

JITServer disabled:

`-XX:+EnableCRIUSupport`
`-XX:+EnableCRIUSupport -XX:-CRIURestoreNonPortableMode`
`-XX:+EnableCRIUSupport -XX:-UseJITServer`
`-XX:+EnableCRIUSupport -XX:-CRIURestoreNonPortableMode -XX:-UseJITServer`
`-XX:+EnableCRIUSupport -XX:+CRIURestoreNonPortableMode -XX:-UseJITServer`

JITServer enabled, remote compilation delayed until after restore:

`-XX:+EnableCRIUSupport -XX:+CRIURestoreNonPortableMode` (JITServer implicitly enabled)
`-XX:+EnableCRIUSupport -XX:+CRIURestoreNonPortableMode -XX:+UseJITServer`

JITServer enabled, remote compilation not delayed:

`-XX:+EnableCRIUSupport -XX:+UseJITServer`
`-XX:+EnableCRIUSupport -XX:-CRIURestoreNonPortableMode -XX:+UseJITServer`

### CODE REVIEW

**Comment by ymanton** [comment]
FYI @vijaysun-omr

**Comment by vijaysun-omr** [comment]
@mpirvu and @dsouzai should probably review this

File: runtime/jcl/common/criu.cpp
Comment (JasonFengJ9) [inline runtime/jcl/common/criu.cpp:68]: This is a JNI implementation that supports java native method `openj9.internal.criu.InternalCRIUSupport.isNonPortableRestoreModeImpl()` which is not in this PR. I think only `J9InternalFunctions` `isNonPortableRestoreMode` is needed as you already implemented. In that case (java method not required), `Java_openj9_internal_criu_InternalCRIUSupport_isNonPortableRestoreModeImpl` related change should be removed.

Comment (ymanton) [reply runtime/jcl/common/criu.cpp]: Agreed, I've dropped this patch.

**Review by mpirvu — COMMENTED** [review]
Changes look good to me provided `isNonPortableRestoreMode()` returns `false` after the restore operation

**Comment by ymanton** [comment]
> Changes look good to me provided `isNonPortableRestoreMode()` returns `false` after the restore operation

It doesn't, it stays `true` for the life of the process.

`isCheckpointAllowed()` will transition from `true` to `false` after the first restore if `isNonPortableRestoreMode()` is `true`.
`isCheckpointAllowed()` will return `true` for the life of the process if `isNonPortableRestoreMode()` is `false`.

**Review by mpirvu — APPROVED** [review]
ok. That will work.

**Comment by dsouzai** [comment]
Looks good to me. I just had a question about the general idea; in the non portable restore run when the JVM implicitly runs in client mode, how would it know where a JITServer instance would be? I'm guessing it would require some command line options right? Or are there env vars that can be set on the restore run for that purpose? 

The reason I ask is because if the user needs to specify command line options for the server's IP address and what not, then wouldn't they also add `-XX:+UseJITServer`?

**Comment by mpirvu** [comment]
You could use `-XX:JITServerAddress=` without `-XX:+UseJITServer`, though the general question remains: what if I want to attach to a server whose address is not known at checkpoint time.
Agreed that we need some env vars related to setting options during a restore operation. Note that not all options can be changed once the JVM has started because this could create some inconsistencies.

File: runtime/compiler/control/J9Options.cpp
Comment (JasonFengJ9) [inline runtime/compiler/control/J9Options.cpp:2086]: Suggest to share `vm->internalVMFunctions` via a local variable.

**Review by JasonFengJ9 — APPROVED** [review]
VM change looks good.

**Comment by ymanton** [comment]
By default clients look for a JITServer at `localhost:38400`. If you want to change that you need to add command line options and then yes you could just add `-XX:+UseJITServer` at that point, so this implicit enablement is really only a convenience if you accept the default host/port.

As far as affecting the host/port at restore time goes, environment variables are currently "baked into" the process at checkpoint time in the same way that command line options are[^1] so you can't use them to change the host or port either. We don't have a general way to change the command line options or env vars at restore time currently.

I think the most likely scenario is that users are deploying in containers and can use port mapping to redirect `host:port` to wherever they want. If they're deploying directly on host they can set up a global firewall rule or a network namespace and a local firewall rule to do the redirection. If they're doing this kind of remapping they would probably accept the `localhost:38400` default.

[^1]: Unless you are accessing env vars through the checkpoint-aware Java APIs the VM provides to the application.

File: runtime/compiler/control/J9Options.cpp
Comment (ymanton) [reply runtime/compiler/control/J9Options.cpp]: Fixed, thanks.

**Comment by mpirvu** [comment]
> use port mapping to redirect host:port to wherever they want. 

@ymanton Is it possible to redirect "localhost" from inside the container running the client to the IP address of the server?

**Comment by vijaysun-omr** [comment]
Jenkins test sanity all jdk17

**Review by vijaysun-omr — APPROVED** [review]

**Comment by vijaysun-omr** [comment]
I'll wait for @dsouzai to approve formally and this conversation  https://github.com/eclipse-openj9/openj9/pull/15674#issuecomment-1208558380 to end before merging.

**Review by dsouzai — APPROVED** [review]

**Comment by ymanton** [comment]
> @ymanton Is it possible to redirect "localhost" from inside the container running the client to the IP address of the server?

I just tried it to confirm but had no luck. It seems the container's loopback interface is not as easy to affect as the hosts. Packets sent over the host's localhost can be routed via iptables rules, but I couldn't get it to work for the container's localhost, which I had to go through some hoops to expose first as Docker hides it.

The easiest method that I could get to work was using `-XX:JITServerAddress=jitserver` at checkpoint time and when starting the container of the restored process use `docker run --add-host=jitserver:x.x.x.x ...` to set the IP address of `jitserver`. `--add-host` works for both Podman and Docker.

**Pull request merged**
